torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -322,26 +350,103 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
   }
 
-  // TODO: Some ops have conversion happen at Peephole pass.
-  //       The conversion here is incomplete for these ops.
-  //       e.g: ListConstruct, ListUnpack, etc.
-  onnx::ModelProto model_proto;
-  ConvertGraphToONNXProto(n_graph, model_proto, opset_version);
-  GRAPH_DEBUG("ONNX graph to run shape inference: ", prettyPrint(model_proto));
-
-  // infer shape
-  onnx::shape_inference::InferShapes(model_proto);
-  GRAPH_DEBUG("ONNX graph after shape inference: ", prettyPrint(model_proto));
-
-  UpdateOutputTypeByONNXProto(n, clone_node, model_proto);
+  SpecialPostProcess(n);
   GRAPH_DEBUG(
       "Torch graph after shape inference:", n->owningGraph()->toString());
 }
 
+std::shared_ptr<Graph> ONNXSetDynamicInputShape(
+    std::shared_ptr<Graph>& graph,
+    const std::unordered_map<
+        std::string,
+        std::unordered_map<int64_t, std::string>>& dynamic_axes,
+    const std::vector<std::string>& input_names) {
+  GRAPH_UPDATE("ONNX set dynamic input shape.");
+  auto new_graph = graph->copy();
+  GRAPH_UPDATE("dynamic axes tensor names:", [&]() {
+    std::vector<std::string> res;
+    for (auto it : dynamic_axes) {
+      res.emplace_back(it.first);
+    }
+    return res;
+  }());
+
+  std::unordered_map<std::string, ::c10::ShapeSymbol> name_to_sym;
+
+  for (int i = 0; i < input_names.size(); ++i) {
+    auto input_name = input_names[i];
+    if (dynamic_axes.find(input_name) != dynamic_axes.end()) {
+      auto axes_names = dynamic_axes.find(input_name)->second;
+      TORCH_INTERNAL_ASSERT(i < new_graph->inputs().size());
+      auto input_tensor_type =
+          new_graph->inputs()[i]->type()->cast<TensorType>();
+      if (!input_tensor_type) {
+        continue;
+      }
+
+      auto shape = input_tensor_type->symbolic_sizes().sizes().value();
+
+      for (auto pair : axes_names) {
+        auto axis = pair.first;
+        auto name = pair.second;
+        if (name_to_sym.find(name) == name_to_sym.end()) {
+          name_to_sym[name] = ::c10::ShapeSymbol::newSymbol();
+        }
+        shape[axis] = name_to_sym[name];
+      }
+
+      new_graph->inputs()[i]->setType(
+          input_tensor_type->withSymbolicShapes(::c10::SymbolicShape(shape)));
+    }
+  }
+
+  return new_graph;
+}
+
+std::shared_ptr<Graph> ONNXAssignOutputShape(
+    std::shared_ptr<Graph>& graph,
+    at::ArrayRef<at::Tensor> outputs,
+    bool onnx_shape_inference) {
+  auto retval = graph->copy();
+  TORCH_INTERNAL_ASSERT(retval->outputs().size() == outputs.size());
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    if (onnx_shape_inference) {
+      retval->outputs()[i]->setType(MergeInferredType(
+          TensorType::create(outputs[i]), retval->outputs()[i]->type()));
+    } else {
+      retval->outputs()[i]->inferTypeFrom(outputs[i]);
+    }
+  }
+  return retval;
+}
+
+std::shared_ptr<Graph> ONNXShapeTypeInference(
+    std::shared_ptr<Graph>& graph,
+    int opset_version) {
+  auto retval = graph->copy();
```

Comment:
Is this is a deep copy? I am wondering if we need to deep copy every time we update the graph with propagate shape info. I see this above in `ONNXSetDynamicInputShape` and `ONNXAssignOutputShape` as well.

Reply:
yea, i think we can update on the original graph, will make the changes.